### DISCUSSION TITLE
Python bindings

### DISCUSSION THREAD

**Discussion by johnnybigert** [elalish/manifold#435] [Q&A]
Hi,

First of all, thank you for what seems to be a fantastic mesh library!

I'm trying to evaluate Manifold from Python, but I'm struggling with some basics. I managed to create a couple of Manifold instances and apply union, intersection, diff etc. I did it manually by reading an OFF file into vertices and indices and then creating a Mesh instance. However, I feel there should be an easier way to create the Manifold directly from the OFF file, but I can't find any import/export functions in the Python bindings and its documentation. Are there any? 

I was looking through the bindings C code and found plenty of import/export functionality, but I fail to see how I could use them from Python. I understand you have GLTF export capabilities, and this is something I would really like to use. Thanks!

**Comment by elalish** [comment]
Thank you! So, from a maintainer perspective, I'd really like to avoid being in the file format support business, since other libraries do that much better. I want to keep our library small so that CAD projects (which already have their own file format support) can easily use us without bringing dependencies they don't want. 

@geoffder convinced me to link our very basic `meshIO` part, built on `Assimp`, into our C bindings (optionally). However, for JS and Python, I think it's much better to use an existing package in those languages for file IO. I've been putting some effort into glTF support on the JS side, using `gltf-transform` (external to our WASM, but here for people to use). It would be great to have similar glue code for a good Python library.

The other problem is that most file formats don't actually represent manifolds, so most of what you read in is unusable. I'd recommend focusing on 3MF (or glTF with my new `EXT_manifold` extension), since those are the only formats that make an effort to represent manifold meshes. 

Our Python side could definitely use some love, so contributions would be most welcome! I'd really like to publish an official package to pypi.

**Comment by geoffder** [comment]
Since the C bindings expose Manifold in a readily usable way to many obscure languages (like ocaml) that don't boast the ecosystems of python/js, I thought it would be helpful to provide IO batteries to spare users from having to write serialization/deserialization or more library bindings for their language just to get off the ground.

That said, you may find this [package](https://pypi.org/project/pygltflib/) useful if you'd like to work with glb/gltf from python.

**Reply by johnnybigert** [reply]
Thanks for the pointer to the gltf package, I managed to make it work and exported a GLTF.
